_posts/2018-01-07-building-serverless-radiator.md

```diff
@@ -0,0 +1,225 @@
+    [:html
+     [:head
+      [:meta {:charset    "UTF-8"
+              :http-equiv "refresh" :content "10"}]
+      [:meta {:http-equiv "Cache-Control" :content= "no-store"}]
+      [:title config/title]
+      styles]
+     [:body
+      page-header
+      (content)]]))
+```
+One limitation of the serverless approach is that all data for rendering the page should be returned in one response or be linked to outside resources. Lambda is not a web server so there is no way to serve static resources like CSS files. One could of course use S3 or another approach for static files.
+
+### Mix some styles and images
+
+Instead of linking to CSS files I chose to embed necessary CSS in HTML response. For this I used [Garden](https://github.com/noprompt/garden) library which makes CSS generation quite similar like using Hiccup.
+
+The next code snippet shows part of the CSS for the radiator.
+
+```clojure
+(def styles
+  [:style
+    (css [:body {:margin  "0px"
+                 :padding "0px"}])
+    (css [:h1 {:font-family    "proxima-nova,sans-serif"
+               :text-transform "uppercase"
+               :margin         "0px"
+               :font-size      "18px"}])])
+```
+
+The Radiator needs of course some fancy images to visually show when something has gone wrong - alert icons etc. Because HTML images are simply links one can link images from anywhere in Internet. GitHub version of the radiator have preconfigured links to Tango Desktop icons which work quite nicely. The images can be changed via configuration.
+
+### Configuring endpoints
+
+The configuration of the endpoints is done with a simple data structure.
+
+```clojure
+(def projects
+  [{:name             "example account"
+    :aws              {:uri     "https://uri-to.com/api-gateway"
+                       :api-key "secret-api-key"}
+    }])
+```
+
+### Supporting different CI providers
+
+The Radiator should of course support also another CI providers like [GitLab](https://about.gitlab.com/). To make this easier I made a generic data structure for pipelines, alarms and metrics. The radiator should therefore first fetch data from endpoints, after that transform them to generic format which is then rendered to Radiator view.
+
+The generic data structure is defined as Clojure Spec as next with example:
+
+```clojure
+(s/def ::name string?)
+(s/def ::pipeline-status #{:success :in-progress :failed :unknown})
+(s/def ::pipeline (s/keys :req-un [::name ::pipeline-status]))
+(s/def ::alarm-status #{:ok :alarm})
+(s/def ::alarm (s/keys :req-un [::name ::alarm-status]))
+(s/def ::metric (s/keys :req-un [::name ::metric-value]))
+
+;; Pipeline example
+{:name "InfraBuild"
+:pipeline-status :in-progress}
+
+;; Alarm example
+{:name "Service-500-alarm"
+:alarm-status :ok}
+
+;; Metric example
+{:name "RDS avg. CPU"
+:metric-value 1.3}
+```
+
+Supporting the GitLab is done simply fetching pipeline data via REST api and after that transforming a pipeline status to generic format.
```

Comment:
"Supporting GitLab is done simply *by* fetching pipeline data via *a* REST *API* and…"